### ISSUE TITLE
Gateway hosts does not work

### ISSUE THREAD

**Issue by dimakievua** [istio/istio#12655]
**Describe the bug**

Try to limit istio gateway to ser ve only requests from one host:

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: bookinfo-gateway
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - fuck.this.dumn.host
    port:
      name: http
      number: 80
      protocol: HTTP
```
Try to test and found it absolutelly has no effect. If I put there any domain name or * it pass everything and block only on VirtualService level:

> 
> $> curl -v -s --head --header 'Host:fuck.this.dumn.host'  http://internal-xxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxx.elb.amazonaws.com/productpage
> * About to connect() to internal-xxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxx.elb.amazonaws.com port 80 (#0)
> *   Trying 10.25.197.162...
> * Connected to internal-xxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxx.elb.amazonaws.com (10.25.197.162) port 80 (#0)
> > HEAD /productpage HTTP/1.1
> > User-Agent: curl/7.29.0
> > Accept: */*
> > Host:fuck.this.dumn.host
> > 
> < HTTP/1.1 200 OK
> HTTP/1.1 200 OK
> < content-type: text/html; charset=utf-8
> content-type: text/html; charset=utf-8
> < content-length: 5723
> content-length: 5723
> < server: envoy
> server: envoy
> < date: Thu, 21 Mar 2019 09:25:27 GMT
> date: Thu, 21 Mar 2019 09:25:27 GMT
> < x-envoy-upstream-service-time: 24
> x-envoy-upstream-service-time: 24
> 
> < 
> * Connection #0 to host internal-xxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxx.elb.amazonaws.com left intact
> $> curl -v -s --head --header 'Host:fuck.it.all'  http://internal-xxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxx.elb.amazonaws.com/productpage
> * About to connect() to internal-xxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxx.elb.amazonaws.com port 80 (#0)
> *   Trying 10.25.235.109...
> * Connected to internal-xxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxx.elb.amazonaws.com (10.25.235.109) port 80 (#0)
> > HEAD /productpage HTTP/1.1
> > User-Agent: curl/7.29.0
> > Accept: */*
> > Host:fuck.it.all
> > 
> < HTTP/1.1 200 OK
> HTTP/1.1 200 OK
> < content-type: text/html; charset=utf-8
> content-type: text/html; charset=utf-8
> < content-length: 5719
> content-length: 5719
> < server: envoy
> server: envoy
> < date: Thu, 21 Mar 2019 09:25:38 GMT
> date: Thu, 21 Mar 2019 09:25:38 GMT
> < x-envoy-upstream-service-time: 25
> x-envoy-upstream-service-time: 25
> 
> < 
> * Connection #0 to host internal-xxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxx.elb.amazonaws.com left intact

**Expected behavior**
I tried to find good explanation what exacly hosts in istio gateway should do but didn't find it. 
My understanding it should pass only request from mentioned host, but block all other.

**Steps to reproduce the bug**
Just create istio gateway with any URL in hosts.

**Version**

> $> istioctl version -c ~/.kube/kubeconfig
> Version: 1.0.5
> GitRevision: c1707e45e71c75d74bf3a5dec8c7086f32f32fad
> User: root@6f6ea1061f2b
> Hub: docker.io/istio
> GolangVersion: go1.10.4
> BuildStatus: Clean
> 
> $> kubectl version
> Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.8-eks-7c34c0", GitCommit:"7c34c0d2f2d0f11f397d55a46945193a0e22d8f3", GitTreeState:"clean", BuildDate:"2019-03-01T22:49:39Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
> 

**Installation**
With Helm

**Environment**
EKS

**Cluster state**

**Comment by rshriram** [comment]
where is the virtual service? gateway by itself is not sufficient

**Comment by dimakievua** [comment]
Hi, this is standard bookinfo virtualservice from istio tutorial:

```
Name:         bookinfo
Namespace:    default
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"networking.istio.io/v1alpha3","kind":"VirtualService","metadata":{"annotations":{},"name":"bookinfo","namespace":"default"}...
API Version:  networking.istio.io/v1alpha3
Kind:         VirtualService
Metadata:
  Creation Timestamp:  2019-02-20T17:20:04Z
  Generation:          1
  Resource Version:    8430
  Self Link:           /apis/networking.istio.io/v1alpha3/namespaces/default/virtualservices/bookinfo
  UID:                 c354571e-3533-11e9-b928-0af271e79d40
Spec:
  Gateways:
    bookinfo-gateway
  Hosts:
    *
  Http:
    Match:
      Uri:
        Exact:  /productpage
      Uri:
        Exact:  /login
      Uri:
        Exact:  /logout
      Uri:
        Prefix:  /api/v1/products
    Route:
      Destination:
        Host:  productpage
        Port:
          Number:  9080
Events:            <none>
```

**Comment by dimakievua** [comment]
Actually if I set URL in virtualservice's hosts it successfully filter by it. So I thought in Gateway it should be same behavior.

**Comment by linsun** [comment]
do you mean if you specify `Hosts: fuck.this.dumn.host` in the VS resource, it works fine now?

**Comment by linsun** [comment]
i think it is a bug either way.  either we fail to validate host: * is not valid in VS resource or we fail to bind your VS that has host: * to the corresponding gateway.  

I'll defer to @rshriram @frankbu @louiscryan for which direction we should go.

**Comment by rshriram** [comment]
* is valid in virtual service if it binds to a gateway..
Restricting at the gateway hosts alone is not sufficient. It is a known bug in the system that if you have a unrestricted host in vs but not in gateawy, the one in vs takes over. So I would suggest that you use the same hosts in both

**Comment by linsun** [comment]
i think this is also fixed by #12663

**Comment by frankbu** [comment]
@linsun, no, this not fixed by #12663. It is a very different problem. 

As @rshriram mentioned, this is a known issue when using `*` in the virtualservice host. We should document this for now, and hopefully fix it in the near future.

**Comment by dimakievua** [comment]
Hi. If issue will be documented this you will avoid questions regarding this. Thank you for help.

**Comment by frankbu** [comment]
This issue is fixed by #12847. The more restrictive host in the gateway will be honored.
